bundles/event_admin/event_admin/src/celix_event_admin.c

```diff
@@ -415,6 +454,131 @@ int celix_eventAdmin_removeEventHandlerWithProperties(void* handle, void* svc, c
+        while (!celix_stringHashMapIterator_isEnd(&iter)) {
+            celix_event_seq_id_cache_t* cache = iter.value.ptrValue;
+            if (celix_elapsedtime(CLOCK_MONOTONIC, cache->lastModified) > 60*60/*1h*/) {
+                celix_stringHashMapIterator_remove(&iter);
+            } else {
+                celix_stringHashMapIterator_next(&iter);
+            }
+        }
+    }
+    return;
+}
+
+static bool celix_eventAdmin_isDuplicateEvent(celix_event_admin_t* ea, const char* topic CELIX_UNUSED, const celix_properties_t* properties) {
+    const char* remoteFwUUID = celix_properties_get(properties, CELIX_EVENT_REMOTE_FRAMEWORK_UUID, NULL);
+    if (remoteFwUUID == NULL) {
+        return false;
+    }
+    long seqId = celix_properties_getAsLong(properties, CELIX_EVENT_REMOTE_SEQ_ID, -1L);
+    if (seqId <= 0) {
+        return false;
+    }
+    long seqIdMod = seqId % CELIX_EVENT_ADMIN_MAX_EVENT_SEQ_ID_CACHE_SIZE;
+    celix_auto(celix_rwlock_wlock_guard_t) wLockGuard = celixRwlockWlockGuard_init(&ea->lock);
+    celix_event_seq_id_cache_t* seqIdCache = celix_stringHashMap_get(ea->eventSeqIdCache, remoteFwUUID);
+    if (seqIdCache == NULL) {
+        celix_autofree celix_event_seq_id_cache_t* cache = calloc(1, sizeof(*cache));
+        if (cache == NULL) {
+            celix_logHelper_error(ea->logHelper, "Failed to create event seq id cache for %s.", remoteFwUUID);
+            return false;
+        }
+        celix_status_t status = celix_stringHashMap_put(ea->eventSeqIdCache, remoteFwUUID, cache);
+        if (status != CELIX_SUCCESS) {
+            celix_logHelper_error(ea->logHelper, "Failed to add event seq id cache for %s.", remoteFwUUID);
+            return false;
+        }
+        seqIdCache = celix_steal_ptr(cache);
+    }
+    seqIdCache->lastModified = celix_gettime(CLOCK_MONOTONIC);
+    if (seqIdCache->seqIdBuffer[seqIdMod] == seqId) {
+        return true;
+    }
+    seqIdCache->seqIdBuffer[seqIdMod] = seqId;
+
+    celix_eventAdmin_retrieveLongTimeUnusedEventSeqIdCache(ea);
+
+    return false;
+}
+
+static long celix_eventAdmin_getEventSeqId(celix_event_admin_t* ea) {
+    celix_auto(celix_rwlock_wlock_guard_t) wLockGuard = celixRwlockWlockGuard_init(&ea->lock);
+    long seqId = ea->nextSeqId++;
+    if (seqId <= 0) {
+        seqId = 1;
+        ea->nextSeqId = seqId + 1;
+    }
+    return seqId;
+}
+
+static int celix_eventAdmin_deliverEventToRemote(celix_event_admin_t* ea, const char* topic, const celix_properties_t* props, bool async) {
+    celix_autoptr(celix_properties_t) remoteProps = celix_properties_copy(props);
+    if (remoteProps == NULL) {
+        celix_logHelper_error(ea->logHelper, "Failed to copy remote properties for event %s.", topic);
+        return ENOMEM;
+    }
+    celix_properties_unset(remoteProps, CELIX_EVENT_REMOTE_ENABLE);
+    celix_status_t status = celix_properties_set(remoteProps, CELIX_EVENT_REMOTE_FRAMEWORK_UUID, ea->fwUUID);
```

Comment:
👍 I like the design of an opt-in with CELIX_EVENT_REMOTE_ENABLE and the "this is a remote event" indication with CELIX_EVENT_REMOTE_FRAMEWORK_UUID.